File: firmware/components/protocol/protocol_frame.c
#include "protocol_frame.h"

static void write_u16le(uint8_t *buf, uint16_t v) {
    buf[0] = (uint8_t)(v & 0xFF);
    buf[1] = (uint8_t)((v >> 8) & 0xFF);
}

static uint16_t read_u16le(const uint8_t *buf) {
    return (uint16_t)(buf[0] | ((uint16_t)buf[1] << 8));
}

static void write_u32le(uint8_t *buf, uint32_t v) {
    buf[0] = (uint8_t)(v & 0xFF);
    buf[1] = (uint8_t)((v >> 8) & 0xFF);
    buf[2] = (uint8_t)((v >> 16) & 0xFF);
    buf[3] = (uint8_t)((v >> 24) & 0xFF);
}

static uint32_t read_u32le(const uint8_t *buf) {
    return (uint32_t)buf[0] | ((uint32_t)buf[1] << 8) |
           ((uint32_t)buf[2] << 16) | ((uint32_t)buf[3] << 24);
}

size_t encode_beacon_frame(const beacon_frame_t *in, uint8_t *out_buf) {
    out_buf[0] = FRAME_TYPE_BEACON;
    write_u16le(out_buf + 1, in->sender_id);
    out_buf[3] = in->hop_count;
    return BEACON_FRAME_LEN;
}

size_t encode_join_frame(const join_frame_t *in, uint8_t *out_buf) {
    out_buf[0] = FRAME_TYPE_JOIN;
    write_u16le(out_buf + 1, in->sender_id);
    write_u16le(out_buf + 3, in->target_parent_id);
    return JOIN_FRAME_LEN;
}

size_t encode_data_frame(const data_frame_t *in, uint8_t *out_buf) {
    out_buf[0] = FRAME_TYPE_DATA;
    write_u16le(out_buf + 1, in->sender_id);
    out_buf[3] = (uint8_t)in->needs_water;
    out_buf[4] = in->battery_pct;
    write_u32le(out_buf + 5, in->timestamp);
    return DATA_FRAME_LEN;
}

int decode_frame_type(const uint8_t *buf, size_t len) {
    if (len == 0) {
        return -1;
    }
    return buf[0];
}

int decode_beacon_frame(const uint8_t *buf, size_t len, beacon_frame_t *out) {
    if (len != BEACON_FRAME_LEN || buf[0] != FRAME_TYPE_BEACON) {
        return -1;
    }
    out->sender_id = read_u16le(buf + 1);
    out->hop_count = buf[3];
    return 0;
}

int decode_join_frame(const uint8_t *buf, size_t len, join_frame_t *out) {
    if (len != JOIN_FRAME_LEN || buf[0] != FRAME_TYPE_JOIN) {
        return -1;
    }
    out->sender_id = read_u16le(buf + 1);
    out->target_parent_id = read_u16le(buf + 3);
    return 0;
}

int decode_data_frame(const uint8_t *buf, size_t len, data_frame_t *out) {
    if (len != DATA_FRAME_LEN || buf[0] != FRAME_TYPE_DATA) {
        return -1;
    }
    out->sender_id = read_u16le(buf + 1);
    out->needs_water = (needs_water_t)buf[3];
    out->battery_pct = buf[4];
    out->timestamp = read_u32le(buf + 5);
    return 0;
}

File: hub/protocol_frame.py
"""Python port of firmware/components/protocol/protocol_frame.c.

Must stay byte-for-byte consistent with the C version: any change to one
requires the same change in the other (see the M3 plan's Task 2 note). The
frame-type constants, field order, field widths, and little-endian encoding
below are a direct mirror of protocol_frame.h/.c.

The C API returns an int status code through an out-param; here decode
returns the frame dataclass directly, or None on a type/length mismatch —
the byte layout is what must match, not the calling convention.
"""

from dataclasses import dataclass
from enum import IntEnum
from typing import Optional, Union

FRAME_TYPE_BEACON = 0x01
FRAME_TYPE_JOIN = 0x02
FRAME_TYPE_DATA = 0x03

BEACON_FRAME_LEN = 4
JOIN_FRAME_LEN = 5
DATA_FRAME_LEN = 9
PROTOCOL_FRAME_MAX_LEN = DATA_FRAME_LEN


class NeedsWater(IntEnum):
    FALSE = 0
    TRUE = 1
    NONE = 2


@dataclass
class BeaconFrame:
    sender_id: int
    hop_count: int


@dataclass
class JoinFrame:
    sender_id: int
    target_parent_id: int


@dataclass
class DataFrame:
    sender_id: int
    needs_water: NeedsWater
    battery_pct: int
    timestamp: int


Frame = Union[BeaconFrame, JoinFrame, DataFrame]


def encode_beacon_frame(frame: BeaconFrame) -> bytes:
    return (
        bytes([FRAME_TYPE_BEACON])
        + frame.sender_id.to_bytes(2, "little")
        + bytes([frame.hop_count])
    )


def encode_join_frame(frame: JoinFrame) -> bytes:
    return (
        bytes([FRAME_TYPE_JOIN])
        + frame.sender_id.to_bytes(2, "little")
        + frame.target_parent_id.to_bytes(2, "little")
    )


def encode_data_frame(frame: DataFrame) -> bytes:
    return (
        bytes([FRAME_TYPE_DATA])
        + frame.sender_id.to_bytes(2, "little")
        + bytes([int(frame.needs_water)])
        + bytes([frame.battery_pct])
        + frame.timestamp.to_bytes(4, "little")
    )


def decode_frame_type(buf: bytes) -> Optional[int]:
    if len(buf) == 0:
        return None
    return buf[0]


def decode_beacon_frame(buf: bytes) -> Optional[BeaconFrame]:
    if len(buf) != BEACON_FRAME_LEN or buf[0] != FRAME_TYPE_BEACON:
        return None
    return BeaconFrame(
        sender_id=int.from_bytes(buf[1:3], "little"),
        hop_count=buf[3],
    )


def decode_join_frame(buf: bytes) -> Optional[JoinFrame]:
    if len(buf) != JOIN_FRAME_LEN or buf[0] != FRAME_TYPE_JOIN:
        return None
    return JoinFrame(
        sender_id=int.from_bytes(buf[1:3], "little"),
        target_parent_id=int.from_bytes(buf[3:5], "little"),
    )


def decode_data_frame(buf: bytes) -> Optional[DataFrame]:
    if len(buf) != DATA_FRAME_LEN or buf[0] != FRAME_TYPE_DATA:
        return None
    return DataFrame(
        sender_id=int.from_bytes(buf[1:3], "little"),
        needs_water=NeedsWater(buf[3]),
        battery_pct=buf[4],
        timestamp=int.from_bytes(buf[5:9], "little"),
    )
